### ISSUE TITLE
Use treat instead of anydata

### ISSUE THREAD

**Issue by boriborm** [pljson/pljson#62]
On my test server i replace using _sys.anydata_ in _json_value_, _json_ and _json_list_ object to use superobject _json_object_ for _json_ and _json_list_ and use _treat_ in json_printer.

Server for test database slower then server for production database, but result for this code

```
declare
 j json;
 jl json_list;
 j2 json;
 cnt number:=5000; --count json objects in json list. Length of all json ~ 275k chars
 start_procedure number;
 end_procedure number;
begin
  jl:=json_list();
  for i in 1..cnt loop
        j2:=json('{p1:"1234567890abcdefghij", p2:"1234567890abcdefghij"}'); -- length 54 chars
        jl.append(j2.to_json_value);
  end loop;
  j:=json();
  j.put('data', jl.to_json_value);
  j.put('succes',true);      
  dbms_output.put_line('count = '||cnt);
  start_procedure:=dbms_utility.get_time;
  j.print;
  end_procedure:=dbms_utility.get_time;
  dbms_output.put_line('time '||((end_procedure-start_procedure)/100));
end;
```

on production database:
execution time ~6.5 seconds, print execution time ~3.5 seconds

on test database with changes:
execution time ~3,5 seconds, print execution time ~1 second.

other test database on the same server (same physical resources) where is test database with changes:
execution time ~8 seconds, print execution time ~4.5 seconds

I can rewrite code of library to use treat, but treat work from version Oracle 9.2(?). Is it problem for backward compatibility?

**Comment by boriborm** [comment]
@dsnz . I make fork of develop branch and change code. You can see it and test on boriborm/develop
[https://github.com/boriborm/pljson/tree/develop](url)

My latest code for test new changes

```
declare
 j json;
 jl json_list;
 j2 json;
 cnt number:=10000;
 start_procedure number;
 end_procedure number;
 my_clob clob;
begin
  start_procedure:=dbms_utility.get_time;

  jl:=json_list();
  for i in 1..cnt loop
        j2:=json();
        j2.put('index',i);
        j2.put('p1','1234567890abcdefghij');
        j2.put('p2','1234567890abcdefghij');
        jl.append(j2.to_json_value);
  end loop;
  j:=json();
  j.put('data', jl.to_json_value);
  j.put('succes',true);
  end_procedure:=dbms_utility.get_time;
  dbms_output.put_line('create time '||((end_procedure-start_procedure)/100));
      
  dbms_output.put_line('count = '||cnt);
  dbms_lob.createtemporary(my_clob, true);
  start_procedure:=dbms_utility.get_time;
  j.to_clob(my_clob);
  end_procedure:=dbms_utility.get_time;
   dbms_lob.freetemporary(my_clob);
  dbms_output.put_line('to_clob time '||((end_procedure-start_procedure)/100));
end;
```

On my test database it work ~ 2.5 second after changes,  ~20 seconds - before changes

**Comment by boriborm** [comment]
My Oracle DBA says all test databases on the same physical server :)

**Comment by dsnz** [comment]
Then I guess I will have to spend the weekend :smile: looking at it and doing my own tests too, though I have a feeling that this change is fine. Allow me some days to respond with my results.

**Comment by jsumners** [comment]
What are your results when parsing data like is discussed in https://github.com/pljson/pljson/issues/61 ?

**Comment by boriborm** [comment]
Use tread not speed up parsing with small amount of objects or arrays, but with big data within him.
Use tread speed up getting and determining objects(json or json_list) in json_value. It used much more in json_printer.

**Comment by Pazus** [comment]
Boriborm, can you make your fork public or grant me access, I'm very interested in your implementation.

**Comment by boriborm** [comment]
Hi, James

You can see changes in fork on
https://github.com/boriborm/pljson/tree/develop/src


С уважением,
Бородулин Максим
icq: 219336745

Блог: http://absinversion.blogspot.com

2016-07-07 9:26 GMT+03:00 Pazus <notifications@github.com>:

**Comment by jsumners** [comment]
How does this affect persistent storage of pljson types in the database? The `anydata` type allows this to be done. And, as @dsnz asked, have you run the test suite with these changes?

**Comment by boriborm** [comment]
All object (json, json_list) can be stored in tables of database. But, if object, used now in tables, then upgrade object types with my changes can make problem with getting this objects from tables :( 
We not store objects in tables, i can't test it.

**Comment by dsnz** [comment]
Hi all,

Eventually I did some tests.
Machine is a VM hosting Oracle 12c and happened to be free for single user development by me only.

Under original implementation

- testsuite runs fine in 4.6 sec
- test by Borimor runs in about 129 sec of which ~ 6 sec to create json obkect and 123 sec in to_clob
(interestingly with PLSQL_OPTIMIZE=0,1 got almost same times and with PLSQL_OPTIMIZE=3 and NATIVE code a bit worse times; also there were cases of extra time ~ +20-30 sec when disconnecting/reconnecting and retesting)

    I did a profiling which I attach in anydata.zip (unzip, view anydata.html and then first report) and which confirms majority of time is in ANYDATA.GETOBJECT
- test of issue #61 was modified to run 100 times for meaningful comparisons and total run time was ~ 10 sec
(when single run was about 0.26 sec)

Under new implementation

- installation did not compile immediately and had to compiled by hand but this is understandable as it is a proof of concept and probably not tested for installation
- testsuite runs fine in ~ 4.5 sec
- test by Borimor runs in 2.65 sec of which ~ 1.4 sec to create json and ~ 1.15 sec in to_clob
(the benefit in create is due to the use of 'to_json_value' in the code which is speeded up)

    again I attach a profiling in treat.zip (unzip, view treat.html and then first report)
- test of issue #61 (modified to run 100 times) in again ~ 10 sec

it is obvious from the profiling that with sys.anydata there is some runtime operation done that is costly while with treat it's a simple compile time cast and so there is much less overhead
most probably the problem is "Bug 17487865  Poor performance extracting ANYDATA attribute" which exists until 12.1.0.1 with no indication of a fix from Oracle
I found a mention of this too in original sourceforge discussion list on 2013 but nothing more, this discussion list could not be moved to github of course for ease of access and searching

the benefit with treat will be seen with any operation related to pljson_printer

personally I like much better the use of a parent pljson_object and use of treat,
the full features/benefits of anydata are not needed for this package at all

@boriborm change please any treats to json_list and json to be to **pljson_list** and **pljson** instead and see what is going on with installation

about persistent storage, there is problem only with existing objects stored in tables
perhaps we could do a conversion package

**Comment by jsumners** [comment]
I think the name `pljson_object` as the root is confusing. I read it as being a pljson parsed JSON object. How about `pljson_type` or `pljson_interface` (or something else)?

**Comment by boriborm** [comment]
@jsumners  
pljson_element?
pljson_super? (this is super object for types pljson and pljson_list)

@dsnz. At first i implemented _treat_ in test database (copied from production database) with old pl/json package (version before i began contribute library).  It worked fine. After i downloaded develop branch, implemented changes in files and install on test database. This files I uploaded to my branch develop.

Why old version in production and test?
I used pl/json library in my developed program (external functionality for commercial program).
But with one update of commercial program, developers used old version of pl/json and replaced used newer version with old version (with all errors that we closed on github :) ). After update commercial version usually i update pl/json with latest version, but at last time has not yet done.

Developers of commercial program promised include to update latest version (with new named objects (pljson)) of library :)

**Comment by jsumners** [comment]
I can live with either of those. `pljson_super` may be my preference.

**Comment by jsumners** [comment]
As for it being a breaking change (not being able to retrieve old entries): I'm fine with it. The upcoming release is 2.0.0 which signifies API incompatibility with 1.x.x.

If a tool is to be provided for migrating old data then I think that tool should be in its own repository. We can provide a link to it in the release notes.

**Comment by dsnz** [comment]
@boriborm 
I think it's time to implement the object oriented version instead of sys.anydata
(but keep the 'obj_type' attribute because it is used often)
will you please go on with this ?

I don't think any user stores json objects in database, just json text strings
so risk for backward compatibility is trivial
but speed enhancement is great
and really the proper way to implement a json api is with object inheritance

**Comment by boriborm** [comment]
I lost my job and can't work with Oracle DB now. But version with THREAT in my fork of pljson. Branch DEVELOP. This old version of fork. To merge code need handwork. Keywords for looking changes is pljson_object, threat.

**Comment by jsumners** [comment]
:( sorry to hear that @boriborm 

This would be a good feature to get in before a 2.0.0 release.

**Comment by dsnz** [comment]
yes, sorry to hear indeed !

well I will schedule this enhancement as first to be done in the near future ...

**Comment by dsnz** [comment]
has been implemented in 2.0
